### PR TITLE
micro-http: Add more flexibility when parsing the headers

### PR DESCRIPTION
**Pull request by wt-l00**
## Reason for This PR

Partially addresses #1975 and #1976 

## Description of Changes

Any number of whitespace characters are accepted after ":" when parsing HTTP headers.

- [ ] This functionality can be added in [`micro-http`](https://github.com/firecracker-microvm/firecracker/tree/master/src/micro_http).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

**Comment by wt-l00** [comment]
@aghecenco Thanks for your review! I fixed them.

File: src/micro_http/src/common/headers.rs [hunk 1]
```diff
@@ -472,6 +473,51 @@ mod tests {
             .is_err());
     }
 
+    #[test]
+    fn test_parse_header_in_any_whitespace() {
+        let mut header = Headers::default();
+        // Test valid accecpt any number of whitespace characters.
```
Comment (aghecenco) [inline src/micro_http/src/common/headers.rs:479]:
```suggestion
        // Test that any number of whitespace characters are accepted before the header value.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,7 +27,7 @@
 - Added metrics for the vsock device.
 - Added devtool strip command which removes debug symbols from the release
   binaries.
-
+- Allow no space after ":"parsing header
```
Comment (aghecenco) [inline CHANGELOG.md:30]:
```suggestion
- Any number of whitespace characters are accepted after ":" when parsing HTTP
  headers.
```

File: src/micro_http/src/common/headers.rs [hunk 3]
```diff
@@ -378,9 +378,10 @@ mod tests {
         assert_eq!(headers.content_length, 55);
         assert_eq!(headers.accept, MediaType::ApplicationJson);
 
-        // Valid headers.
+        // Valid headers. (${HEADER_NAME} : WHITESPACE ${HEADER_VALUE})
+        // any number of whitespace characters should be accepted includes zero.
```
Comment (aghecenco) [inline src/micro_http/src/common/headers.rs:382]:
```suggestion
        // Any number of whitespace characters should be accepted including zero.
```

File: src/micro_http/src/common/headers.rs [hunk 4]
```diff
@@ -472,6 +473,51 @@ mod tests {
             .is_err());
     }
 
+    #[test]
+    fn test_parse_header_in_any_whitespace() {
```
Comment (ioanachirca) [inline src/micro_http/src/common/headers.rs:477]: nit:
```suggestion
    fn test_parse_header_whitespaces() {
```

**Comment by ioanachirca** [comment]
> Couple more nits. I'd still prefer if all the commits were squashed into a single one as the changes are small and logically connected.

The commit message could be rephrased to something like this: "Any number of whitespace characters are accepted after ":" when parsing HTTP headers."

File: src/micro_http/src/common/headers.rs (hunk 1 above)
Comment (wt-l00) [reply src/micro_http/src/common/headers.rs]: @aghecenco  Thanks! I fixed it.

File: CHANGELOG.md (hunk 2 above)
Comment (wt-l00) [reply CHANGELOG.md:30]: @lauralt Sorry... I added one more newline.

File: src/micro_http/src/common/headers.rs (hunk 3 above)
Comment (wt-l00) [reply src/micro_http/src/common/headers.rs]: @aghecenco I fixed it.

File: src/micro_http/src/common/headers.rs (hunk 4 above)
Comment (wt-l00) [reply src/micro_http/src/common/headers.rs]: @ioanachirca Thanks for your review! I fixed it.

File: src/micro_http/src/common/headers.rs [hunk 5]
```diff
@@ -472,6 +473,51 @@ mod tests {
...
+
+        // For ContentType
+        assert!(header
+            .parse_header_line(b"Content-Type:application/json")
+            .is_ok());
+        assert!(header
+            .parse_header_line(b"Content-Type:    application/json")
+            .is_ok());
+
+        // For Expect
```
[the remark this answers is not in the document]
Comment (wt-l00) [reply src/micro_http/src/common/headers.rs]: @lauralt I fixed!

File: src/micro_http/src/common/headers.rs [hunk 6]
```diff
@@ -472,6 +473,51 @@ mod tests {
...
+        assert!(header
+            .parse_header_line(b"Transfer-Encoding:    chunked")
+            .is_ok());
+        assert!(header.chunked());
+
+        // For Server
+        assert!(header.parse_header_line(b"Server:xxx.yyy.zzz").is_ok());
+        assert!(header.parse_header_line(b"Server:   xxx.yyy.zzz").is_ok());
+
+        // For Accept
```
Comment (lauralt) [inline src/micro_http/src/common/headers.rs:512]: I might be wrong but I think before the last push we were testing here the Expect case.

Comment (wt-l00) [reply src/micro_http/src/common/headers.rs]: I fixed it!

File: CHANGELOG.md (hunk 2 above)
Comment (ioanachirca) [reply CHANGELOG.md:30]: Please update this line :D

**Comment by aghecenco** [comment]
One more change needed: looks like your PR increases the coverage on AMD (nice!), please update the expected coverage here: https://github.com/firecracker-microvm/firecracker/blob/master/tests/integration_tests/build/test_coverage.py#L26
```diff
diff --git a/tests/integration_tests/build/test_coverage.py b/tests/integration_tests/build/test_coverage.py
index 87ad0e5..6330bc2 100644
--- a/tests/integration_tests/build/test_coverage.py
+++ b/tests/integration_tests/build/test_coverage.py
@@ -23,7 +23,7 @@ import host_tools.proc as proc
 # this contains the frequency while on AMD it does not.
 # Checkout the cpuid crate. In the future other
 # differences may appear.
-COVERAGE_DICT = {"Intel": 84.37, "AMD": 84.32}
+COVERAGE_DICT = {"Intel": 84.37, "AMD": 84.38}
 PROC_MODEL = proc.proc_type()
 
 COVERAGE_MAX_DELTA = 0.05
```

**Comment by wt-l00** [comment]
@aghecenco @lauralt thank you for your help!!

**Pull request merged**